### DISCUSSION TITLE
Fuzzing of a locally integrated analyzer

### DISCUSSION THREAD

**Discussion by predator89090** [zeek/zeek#3862] [Help] [answered]
Hello! 

In my team, we use Zeek as part of our network analysis software. We are locally adding new analyzers to zeek to expand the functionality associated with our product.

Not long ago we came up with the idea of ​​conducting fuzz testing of our analyzers. We decided to use OSS-FUZZ on the libfuzzer engine for this.

But we encountered certain problems, I will try to explain my actions in more detail:

1) Initially, I have a separate zeek repository and a separate plugin repository
2) I clone these repositories into oss-fuzz/projects/zeek
![image](https://github.com/user-attachments/assets/ff048c69-c480-44b6-8c6a-c0d8fffbd820)

3) I made a corpus for my analyzer and added it to the path src/fuzzers/corpora, I also edited the CMakeLists.txt file adding my analyzer there in add_generic_analyzer_fuzz_target() because I want to use a common fuzzer build script.

3) Next, I edit the dockerfile using my local repository, make the build.sh script executable inside the dockerfile
![image](https://github.com/user-attachments/assets/10cbb88b-2381-4832-85cd-f903d5fef104)

4) Next, I adjust the build.sh script and leave only the configuration and zeek build in it, without collecting fuzzers

5) As a result, zeek was assembled inside the image, which I saved and inherited from it when creating an image with zeek + my analyzer, in the build script I placed the configuration of the analyzer based on zeek from the base image and the assembly of fuzzers
![image](https://github.com/user-attachments/assets/40d001de-e5be-40fe-8a9e-04f38b7fbdc0)

6) Then I run python3 infra/helper.py build_fuzzers --sanitizer address zeek --> build.sh worked, the analyzer was successfully configured in zeek, ​​the fuzzer binary file and the corpus were pulled up in oss-fuzz/build/out/zeek
![image](https://github.com/user-attachments/assets/3fad698c-1327-4d62-b009-1b4b454a7b7b)

And then I ran into a problem

I'm trying to run  python3 infra/helper.py run_fuzzer zeek zeek-enip_tcp-fuzzer and encounter this behavior
![image](https://github.com/user-attachments/assets/242be6ec-8798-42ef-80cb-1f14dbe34513)

I found where the problem occurs 
![image](https://github.com/user-attachments/assets/7cce8bc4-492b-4b3b-9e99-89083e236f04)

And this is associated with the analyzer tag, which, as I understand it, should be initialized when creating the analyzer component.

In this regard, I have some questions: 
1) Am I trying to fuzz correctly, or what am I doing is pointless?
2) How could I build a fuzzer, but I can’t run it, because in both cases, judging by the code, an analyzer tag is required

I hope you have time to share your thoughts on this matter and suggest how best to do it, maybe you have already encountered such problems.

**P.S sorry for my English :)**

**Comment by JustinAzoff** [comment, marked as answer]
Hi,

I think plugins may not get loaded when fuzzing in that environment, but you can probably get this working using this option to the zeek `configure` script:

```
--include-plugins=PATHS     paths containing plugins to build directly into Zeek
```

If you point that at your `enip` plugin it will build that into zeek, and the analyzer should exist when fuzzing.

In general the scripts for oss-fuzz are a bit weird, as they want everything to be self contained after it is built, but zeek generally doesn't work that way.  You could also try building and running the fuzzer manually which may work better: https://github.com/zeek/zeek/blob/master/src/fuzzers/README.rst
